✨ [open-formulieren/open-forms#4380] Extra StUF-ZDS payment attributes

stevenbal opened the pull request:
Partially closes open-formulieren/open-forms#4380

sergei-maertens commented on stuf_zds_payments/client.py:
```diff
@@ -0,0 +1,92 @@
+from stuf.stuf_zds.models import StufZDSConfig
+
+
+class ZaakOptions(TypedDict):
```
I think you should be able to subclass `stuf.stuf_zds.client.ZaakOptions` and add your additional keys!

stevenbal replied:
Ah hm, my linter was complaining: `Argument to class must be a base class`, possibly because it doesn't recognize the `stuf` library in my environment in this repo

stevenbal committed:
:sparkles: [open-formulieren/open-forms#4380] Extra StUF-ZDS payment attributes

this feature is implemented in this extension to ensure that it does not affect OF master yet, it will be added to Open Forms itself later on

stevenbal committed:
:white_check_mark: [open-formulieren/open-forms#4380] Add tests for plugin

stevenbal committed:
:green_heart: Fix failing CI build due to setuptools causing issues

stevenbal committed:
:construction_worker: Get tests up and running with docker

stevenbal committed:
:wrench: Only support python >=3.11

stevenbal commented on stuf_zds_payments/registration_variables.py:
```diff
@@ -0,0 +1,59 @@
+@register("payment_amount")
+class PaymentAmount(BaseStaticVariable):
+    name = _("Payment amount")
+    data_type = FormVariableDataTypes.float
```
I noticed this was a `float` instead of a `Decimal` in the Objects API variables registration as well, while it is a `Decimal` on `SubmissionPayment`, will this cause issues?

sergei-maertens replied:
we don't know a decimal type in our own things, and JSON doesn't know decimal either, only `number` (which is int or float). So, just do the same and pretend we didn't see anything :grimacing:

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
-      - uses: actions/checkout@v4
-      - uses: actions/setup-python@v5
+      - name: Checkout Open Forms
+        uses: actions/checkout@v3
```
```suggestion
        uses: actions/checkout@v4
```
v3 uses deprecated node version, see the warnings here: https://github.com/open-formulieren/open-forms-ext-stuf-zds-payments/actions/runs/10144521175?pr=1

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
+          path: open-forms
+
+      - name: Checkout StUF-ZDS payments extension
+        uses: actions/checkout@v3
```
```suggestion
        uses: actions/checkout@v4
```

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
+        uses: maykinmedia/setup-django-backend@v1.1
         with:
-          python-version: ${{ matrix.python }}
+          apt-packages: 'libxml2-dev libxmlsec1-dev libxmlsec1-openssl gettext postgresql-client gdal-bin'
```
Having `libxml2-dev` and `libxmlsec1-dev` present has a nasty side-effect that it breaks the uwsgi binary. They're not needed anymore in latest versions of Open Forms, since we can install the binary wheels these days of `lxml` and `xmlsec`.

See https://github.com/open-formulieren/open-forms/commit/6882bd292b531db8872da103a131dcd2e67e5a6e for the patch in OF itself.

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
 
       - name: Publish coverage report
-        uses: codecov/codecov-action@v4
+        uses: codecov/codecov-action@v3.1.4
```
I've added the `CODECOV_TOKEN` secret, so you can use v4 and provide the secret.

v3 is deprecated because of the deprecated Node versions.

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
+          # PRs result in version 'merge' -> transform that into 'latest'
+          [ "$VERSION" == "merge" ] && VERSION=latest
+
+          echo ::set-output name=tag::${VERSION}
```
`set-output` has been deprecated for a while, instead, use env files, like [here](https://github.com/open-formulieren/open-forms/blob/6c8ac92262a0e7e7bbcff30730e9c4e1e6b74eff/.github/workflows/ci.yml#L262)

Feel free to also update the other existing extensions if they have the same issue (I assume you copied over that CI configuration).

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
+          working-directory: ${{ github.workspace }}/open-forms
+          files: ./coverage-extension.xml
+
+  docker_build:
```
Looking at this, I'm wondering a little bit how useful it is to do this docker build in the first place - it doesn't make sense to publish this image by itself, it makes more sense to install the extra libraries with extensions instead, like we do in an OF build.

So feel free to remove this step - I think it used to be relevant when we did the (unit) testing in Docker, but the checkout + symlink approach replaced that.

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
     steps:
-      - uses: actions/checkout@v4
-      - uses: actions/setup-python@v5
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
```
```suggestion
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
```

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
-          pip install build --upgrade
-          python -m build
-
+          pip install pip setuptools wheel --upgrade
+          python setup.py sdist bdist_wheel
```
does the new way with `build` not work?

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
       - name: Publish a Python distribution to PyPI
         uses: pypa/gh-action-pypi-publish@release/v1
-
+        with:
+          user: __token__
+          password: ${{ secrets.PYPI_TOKEN }}
```
I have set up verified publishing in PyPI, so this is not needed.

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,57 +10,138 @@ on:
     name: Publish package to PyPI
     runs-on: ubuntu-latest
-    needs: tests
-    environment: release
-    permissions:
-      id-token: write
```
these are needed for verified publishing :)

sergei-maertens commented on .github/workflows/code_quality.yml:
```diff
@@ -20,12 +20,14 @@ jobs:
       - uses: actions/setup-python@v5
         with:
-          python-version: '3.10'
+          python-version: '3.11'
```
OF is Python 3.12

```suggestion
          python-version: '3.12'
```
or was there a particular reason you use 3.11?

stevenbal replied:
hmm no, I must have missed this one, I guess I should only support 3.12 and drop 3.11 entirely?

sergei-maertens replied:
yep, OF itself is 3.12 only anyway

sergei-maertens commented on .github/workflows/code_quality.yml:
```diff
@@ -20,12 +20,14 @@ jobs:
       matrix:
-        toxenv: [isort, black, flake8, docs]
+        toxenv: [isort, black, flake8]
+    env:
+      PIP_CONSTRAINT: pip-constraints.txt
```
I've seen the Slack thread about this issue here - which library is causing problems? Since OF itself seems to install just fine on CI?

... it could maybe be because of our dependency cache that it's still working :thinking:

stevenbal replied:
Hmm I tried it again and it works now, it may have been an issue for python3.10/3.11

sergei-maertens commented on Dockerfile:
Not sure if we actually need this, see the earlier remark in `ci.yml`

sergei-maertens commented on README.rst:
```diff
@@ -6,7 +6,7 @@ Welcome to stuf_zds_payments's documentation!
 :Source: https://github.com/maykinmedia/stuf_zds_payments
 :Keywords: ``<keywords>``
-:PythonVersion: 3.10
+:PythonVersion: 3.11
```
```suggestion
:PythonVersion: 3.12
```

sergei-maertens commented on README.rst:
```diff
@@ -30,7 +30,7 @@ Installation
 ------------
 
-* Python 3.10 or above
+* Python 3.11 or above
```
```suggestion
* Python 3.12 or above
```

sergei-maertens commented on pyproject.toml:
```diff
@@ -20,12 +20,11 @@ classifiers = [
     "Operating System :: MacOS",
     "Operating System :: Microsoft :: Windows",
-    "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
```
```suggestion
```

sergei-maertens commented on pyproject.toml:
```diff
@@ -20,12 +20,11 @@ classifiers = [
     "Topic :: Software Development :: Libraries :: Python Modules",
 ]
-requires-python = ">=3.10"
+requires-python = ">=3.11"
```
```suggestion
requires-python = ">=3.12"
```

sergei-maertens commented on tox.ini:
```diff
@@ -9,7 +9,6 @@ skip_missing_interpreters = true
 [gh-actions]
 python =
-    3.10: py310
     3.11: py311
```
think this can be dropped too

sergei-maertens reviewed (changes requested):
Part 1 - this focuses on the non-functionals (CI setup etc.)

sergei-maertens commented on stuf_zds_payments/apps.py:
```diff
@@ -1,5 +1,11 @@
 
 class StufZdsPaymentsConfig(AppConfig):
     name = "stuf_zds_payments"
+    label = "stuf_zds_payments"
```
```suggestion
    label = "registration_stuf_zds_payments"
```
to keep up with the existing pattern of properly namespacing it under the `registrations` module :)

sergei-maertens commented on stuf_zds_payments/client.py:
```diff
@@ -0,0 +1,73 @@
+
+
+class ZaakOptions(ZaakOptions):
+    payment_status_update_xml: dict
```
shouldn't this be a string?

```suggestion
    payment_status_update_xml: str
```

sergei-maertens commented on stuf_zds_payments/client.py:
```diff
@@ -0,0 +1,73 @@
+    payment_status_update_xml: dict
+
+
+class Client(_Client):
```
I think that `create_zaak` also needs to be updated, either here or in the plugin, for the case where registration is deferred until payment is received. In those cases, `set_zaak_payment` is never called.

I've been considering overriding `create_zaak` to call `set_zaak_payment` afterwards, but I don't know how that works with the async StUF calls so it's probably best to include that in the `create_zaak` transaction directly :(

Considering how much is happening in `register_submission` and `create_zaak`, a viable route *may* be to hook into `execute_call` instead and based on the `soap_action` modify the behaviour.

olenitsj commented on stuf_zds_payments/templates/registrations/contrib/stuf_zds_payments/payment_status_update_xml.txt:
```diff
@@ -0,0 +1,8 @@
+    <StUF:extraElement naam="payment_completed">{{ tt_openvariable }} payment_completed {{ tt_closevariable }}</StUF:extraElement>
+    <StUF:extraElement naam="payment_amount">{{ tt_openvariable }} payment_amount {{ tt_closevariable }}</StUF:extraElement>
+    <StUF:extraElement naam="payment_public_order_ids">{{ tt_openvariable }} payment_public_order_ids {{ tt_closevariable }}</StUF:extraElement>
+</StUF:extraElementen>
```
@sergei-maertens I expected to see here one more <StUF:extraElement> containing the PaymentID from ogone

sergei-maertens replied:
good catch - @stevenbal this is also something that we need to address still. Okay to do this in another PR, since we'll need to update our payment plugins for this too and I'd like to treat that as a separate item/topic.

stevenbal replied:
created a separate PR for this https://github.com/open-formulieren/open-forms-ext-stuf-zds-payments/pull/3

sergei-maertens commented on stuf_zds_payments/plugin.py:
```diff
@@ -0,0 +1,94 @@
+
+def get_payment_status_update_text() -> str:
+    return render_to_string(
+        "registrations/contrib/stuf_zds_payments/payment_status_update_xml.txt"
```
since it's XML, can't we stick to this instead?

```suggestion
        "registrations/contrib/stuf_zds_payments/payment_status_update.xml"
```

sergei-maertens commented on stuf_zds_payments/plugin.py:
```diff
@@ -0,0 +1,94 @@
+        label=_("payment status update XML template"),
+        help_text=_(
+            "This template is evaluated with the submission data and the resulting XML "
+            "is sent to the StUF-ZDS with a PATCH to update the payment field."
```
```suggestion
            "This template is evaluated with the submission data and the resulting XML "
            "is included in extraElementen."
```
the template will need to be upated, since I'm concerned that we'd be overwriting other extraElementen - any form fields that are not explicitly mapped to a registration attribute are included in the extra_data (via `get_unmapped_data`)

sergei-maertens commented on stuf_zds_payments/plugin.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@register(PLUGIN_IDENTIFIER)
+class StufZDSPaymentsRegistration(StufZDSRegistration):
```
as remarked in the client, I'm afraid we also need changes to the `register_submission` call.

Perhaps a small refactor to break out a new `get_extra_data` method in Open Forms itself can make our life much simpler, I'm okay with backporting such a refactor to the 2.7.x branch. Then the only thing this subclass would have to override would probably be this new method to include the payment data.

Signature would be:

```py
def get_extra_data(self, submission: Submission) -> dict[str, Any]: ...
```

Looking at `src/stuf/stuf_zds/templates/stuf_zds/soap/includes/extraElementen.xml` and the StUF standard (https://vng-realisatie.github.io/StUF-onderlaag/documenten/Stuf0301.pdf), the values must be strings - essentially extraElementen is a key-value pair. I had first thought we could drop any XML in there, but this is possible only through `aanvullendeElementen` (and that's much more complicated, so let's ignore this).

I think this simplifies the configuration aspect in the UI - instead of giving a template, we should probably just allow them to specify the key + variable to fill out/select in a dropdown (maybe?). I think we can do this with a nested serializer, similar to the `to_emails` of the email registration plugin for the list behaviour.

stevenbal replied:
I added `get_extra_data` in https://github.com/open-formulieren/open-forms/pull/4581 and also used this in `update_payment_status`. I think this should be sufficient? It works both for updating and creating Zaken with the extra attributes

sergei-maertens replied:
merged and backport that PR - seems like that would indeed be the minimal approach :)

sergei-maertens commented on stuf_zds_payments/plugin.py:
```diff
@@ -0,0 +1,94 @@
+            options["payment_status_update_xml"], values
+        )
+
+        with get_client(options) as client:
+            client.set_zaak_payment(
+                submission.registration_result["zaak"],
+                payment_status_update_xml=payment_status_update_xml,
+            )
```
this will probably be a bit more complex since we'd have to repeat all the `extraElementen`, I don't think there's a way to only update/create the changes

stevenbal replied:
I moved from the template to a mapping as you suggested, to rely on the existing behaviour for extraElementen

sergei-maertens commented on stuf_zds_payments/plugin.py:
```diff
@@ -0,0 +1,94 @@
+    match value:
+        case bool():
+            return "true" if value else "false"
+        case list():
+            return " ".join(value)
```
have you looked at `stuf_zds.utils.flatten_data`? if using extra-elementen,  this should already handle dicts/lists, so a list like `["foo", "bar"]` would turn into:

```xml
<extraElement naam="list.0">foo</extraElement>
<extraElement naam="list.1">bar</extraElement>
```

sergei-maertens reviewed (changes requested):
Part 2 - python code/actual implementation :)

stevenbal committed:
:ok_hand: [open-formulieren/open-forms#4380] PR feedback

* related to non functional code (CI etc.)
* moved from XML template to a mapping to define which variables get mapped to which extraElementen key
* implemented extra payment attributes for zaak creation (in case zaak creation is deferred until payment is done)

sergei-maertens commented on .github/workflows/ci.yml:
```diff
@@ -10,40 +10,86 @@ on:
+        with:
+          repository: open-formulieren/open-forms
+          path: open-forms
+          # TODO remove after this is merged
+          ref: feature/4380-stuf-zds-extra-payments-changes
```
self-reminder

sergei-maertens commented on stuf_zds_payments/plugin.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class ZaakPaymentOptionsSerializer(ZaakOptionsSerializer):
+    payment_status_update_mapping = serializers.ListField(
+        child=serializers.ListField(
+            child=serializers.CharField(), min_length=2, max_length=2
+        ),
+        label=_("payment status update variable mapping"),
+        help_text=_(
+            "This mapping is used to map the variable keys to keys used in the XML "
+            "that is sent to StUF-ZDS. Those keys and the values belonging to them in "
+            "the submission data are included in extraElementen."
+        ),
+        required=False,
+    )
```
I wondering (but don't know how well this plays with rjsf) if explicit labels are an option?

```py
class MappingSerializer(serializers.Serializer):
    form_variable = serializers.CharField(...)
    stuf_name = serializers.CharField(...)


class ZaakPaymentOptionsSerializer(...)
    payment_status_update_mapping = MappingSerializer(many=True)
```

I think this would be preferable to reason about named parameters instead of positional arguments.

sergei-maertens commented on stuf_zds_payments/plugin.py:
```diff
@@ -0,0 +1,121 @@
+        when creating the Zaak
+        """
+        data = super().get_extra_data(submission, options)
+        data.update(self.get_extra_payment_variables(submission, options))
+        return data
```
```suggestion
        payment_extra = self.get_extra_payment_variables(submission, options)
        return {**data, **payment_extra}
```
a side-effect free alternative :)

sergei-maertens commented on stuf_zds_payments/tests/test_plugin.py:
```diff
@@ -0,0 +1,201 @@
+class StufZDSPaymentsRegistrationTestCase(OFVCRMixin, TestCase):
+    VCR_TEST_FILES = TESTS_DIR / "data"
+
+    def setUp(self):
+        super().setUp()
```
that's a lot of set up shared between the tests - it looks like you should be able to put it in `setUpTestData` instead of running it for every test?

sergei-maertens commented on stuf_zds_payments/plugin.py:
```diff
@@ -0,0 +1,132 @@
+PLUGIN_IDENTIFIER = "stuf-zds-create-zaak:ext-utrecht"
+
+
+def default_payment_status_update_mapping() -> list[dict[str, str]]:
```
(for when we port this to OF core)

I think that `TypedDict` is more appropriate here - now key/value can be any string, but if you do:

```py
class PaymentVariableMapping(TypedDict):
    form_variable: str
    stuf_name: str
```
you get much more accurate types
